### ISSUE TITLE
Problems with control of water coils in outdoor air stream

### ISSUE THREAD

**Issue by mjwitte** [NREL/EnergyPlus#6148]
Issue overview
--------------
This is related to #5695 which was addressed by #5973.

1. User file has hot water preheat coil in the outdoor air stream with economizer active.  Sometimes the coil controls well and at other times it does not.  If the economizer is turned off, then the coil controls exactly as it should.  I suspect these is a problem with the way that timing of the call to the coil controller.  For OA system controllers, this may need a different approach that what is used for controllers on the main air loop branch.

2. Also, I am not sure that the fix in #5973 is general enough.  It appears to disable lockout for all coils on a given airloop simply if these is a hot water coil in the outdoor air system.  If there are also chilled water coils, it seems this fix could prevent lockout when desired.  This needs review and may require additional variables to track lockout individually by coil or for the coil controller to check the lockout status the way that it is done for electric and gas coils.

3. User file has chilled water pre-cool coil in the outdoor air stream.  With differential dry-bulb economizer, coil operates as expected.  With differential enthalpy, it does not.  See ticket 12003 for more details.

### Details
Some additional details for this issue (if relevant):
 - Version of EnergyPlus 8.7.0
 - helpdesk ticket number 12001 and 12003.

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added EnergyPlusDevSupport\DefectFiles
 - [x] Ticket added to Pivotal for defect (development team task)
 - [x] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

**Comment by rraustad** [comment]
This looks promising. Basically what I did was figure out what the water flow rate needed to be at the same time the water coil is simulated. I also bypassed ControlCompOutput and found the answer using RegulaFalsi. This may be the start of a method to replace ControlCompOutput.

    SimOAComponent(
    ..
    } else if ( SELECT_CASE_var == WaterCoil_SimpleHeat ) { // 'Coil:Heating:Water')
        if ( Sim ) {
            if ( CompIndex > 0 ) {
                CalcHotWaterSimpleCoilWaterFlowRate( CompName, FirstHVACIteration, CompIndex );
                ControllerProps( WaterCoil( CompIndex ).ControllerIndex ).BypassControllerCalc = true;
            }
            SimulateWaterCoilComponents( CompName, FirstHVACIteration, CompIndex );
        }
        OAHeatingCoil = true;

Not sure yet why the air flow goes to 0 at times, but when there is air flow, the coil control right on the set point. Node 26 is the air outlet node of the heating coil in the OA System.

![6148_initialtesting](https://user-images.githubusercontent.com/5778156/27102394-1d7bfe7c-5053-11e7-9dd0-a4a873047d4d.png)

**Comment by rraustad** [comment]
I'm not even sure what this flag (CanBeLockedOutByEcono) was meant to do. It's set up now to be false for air loop water coils and true for OA Sys water coils. Also, user has no control over this flag, it's just set to true or false based on where the coil is located. So if economizer is active, the OA system water coils are off, period. There is no check to see of the OA Controller economizer lockout flag is set to NoLockout. Also some confusion as to which coils should be locked out, air loop coils or OA system coils. We need to include lockout logic, if any, in the new code. I think it's already there, and only to lockout economizer operation (not the coil operation), so there may be nothing else to do. @mjwitte can chime in on what should lock out air loop and/or OA system water coils.

    SolveAirLoopControllers(
    RR - around 120 lines down, about line 2590
    // if the controller can be locked out by the economizer operation and the economizer is active, leave the controller inactive
    if ( AirLoopControlInfo( AirLoopNum ).EconoActive && PrimaryAirSystem( AirLoopNum ).CanBeLockedOutByEcono( AirLoopControlNum ) ) {
        ControllerConvergedFlag = true;
        continue;
    }

**Comment by rraustad** [comment]
See PR #6197. PT #147216271 - #6148 Problems with control of water coils in outdoor air stream

**Comment by mjwitte** [comment]
Certainly seems like the user should have some control over this.  Is there a reason OA coils and central branch coils should be treated differently for lockout?  If so, then maybe the lockout input needs more choices?

**Comment by rraustad** [comment]
I can't think of a good reason to lock out the coils (except on OAT), regardless of location. On the other hand, the economizer can be locked out if heating is active, and this is already done. So why the CanBeLockedOutByEcono flag was added in the first place needs to be answered. Someone may have simply thought that if economizer is active, then it's free cooling and no coils should be operating. In my mind it's the reverse, if coils are operating then economizer should be locked out. @EnergyArchmage or @nealkruis or @ejhw or @shorowit or others may have other insights.

**Comment by EnergyArchmage** [comment]
Economizer logic gets far more complex than OAT.  Recirc conditions, enthalpy, various difference.  And yes it isn't free cooling if the coils are still running.  When economizing, the OA flow rates can lift far above minimum so it gets more expensive to run an OA coil with the larger flows.  I agree it is complicated but all that complexity can't also be in the coils, they need to get signal(s) from the OA economizer controller.

**Comment by rraustad** [comment]
@myoldmopar, I've rearranged calls to controllers so that they are called when the coil is simulated (as opposed to calling all controllers first then calling air loop and OA sys water coils). I have the same results as with my other method using RegulaFalsi (see figure above and PR #6176). How should I push this new method? Create a new branch? I want to see how the performance tests speed is impacted.

![6148-usingcontrollers](https://user-images.githubusercontent.com/5778156/27796401-df1b7fe2-5fd7-11e7-9ebd-0f4709068c0d.png)

**Comment by mjwitte** [comment]
@rraustad This is a major structural change for the controllers.  There was a reason for that architecture way back when, but @wfbuhl is probably the only one who can explain it.  Sounds like we need two separate branches, one for this controller restructure and one for the regula-falsi approach, which don't change anything else.  Then we can evaluate speed and diffs without any other changes clouding the picture.

**Comment by EnergyArchmage** [comment]
I think the original design intent went hand in hand with allowing the controlling node to be somewhere farther downstream than the coil outlet,  to account for fan operation and maybe other devices in between.  This is a big change because it would force the control to always be based on coil outlet.  Note that there is lag in the mixed air setpoint controller's fan temperature rise (which would need to be added by transition sometimes) -- so there could be situations where this is a step backward in setpoint control for users who want to use the original design intent to adjust for fan heat.

**Comment by rraustad** [comment]
Trying now to leave main air loops coils to simulate as before. OA sys coils will now iterate on ManageControllers with the coil model (rather sloppy implementation). Hopefully no user files exist where OA sys water coils don't use the coil air outlet node as the sensed node for the controller.

**Comment by rraustad** [comment]
Just came across this. It appears we are now requiring set point nodes to be at the air outlet node of the water coil. I am testing HX assisted water coils with the new methodology and failed to move the SP.

    ** Warning ** HVACControllers: GetControllerInput: Controller:WaterCoil="CENTRAL COOLING COIL CONTROLLER 1". 
    **   ~~~   **  ..Sensor node not found on water coil air outlet node.
    **   ~~~   **  ..The sensor node may have been placed on a node downstream of the coil or on an airloop outlet node.

    ** Severe  ** HVACControllers: Missing temperature setpoint for controller type=Controller:WaterCoil Name="OA CC CONTROLLER 1"
    **   ~~~   ** Node Referenced (by Controller)=OA HX SEC INLET NODE
    **   ~~~   **   use a Setpoint Manager with Control Variable = "Temperature" to establish a setpoint at the controller sensed node.

**Comment by rraustad** [comment]
Brent, I also question this logic:

    CheckIfNodeSetPointManagedByEMS(
        //       AUTHOR         Brent Griffith
    
    for ( Loop = 1; Loop <= numActuatorsUsed + NumExternalInterfaceActuatorsUsed; ++Loop ) {
        if ( ( SameString( EMSActuatorUsed( Loop ).ComponentTypeName, cComponentTypeName ) ) && ( SameString( EMSActuatorUsed( Loop ).UniqueIDName, cNodeName ) ) && ( SameString( EMSActuatorUsed( Loop ).ControlTypeName, cControlTypeName ) ) ) {
            FoundControl = true;
        }
    }

    if ( ( ! ErrorFlag ) && ( ! FoundControl ) ) ErrorFlag = true;

Shouldn't if be if FoundControl were true then return ErrorFlag = true ??
Also, why is ErrorFlag in the IF logic if it is returned from this function?

**Comment by EnergyArchmage** [comment]
Well the check is for missing setpoints, so if it finds one then that is not and error.  I think ErrorFlag is in/out, it may be changed by the routine.

**Comment by rraustad** [comment]
OK, that makes sense since it's been around since 2009. I was having trouble with HX assisted water coils (1 air loop, 1 OA system). Not sure where to put the SP. If I put it at the outlet of the HX (coil system), where I think most users would put the SP, I get an error. It's running now with SP's at outlet of water coil.

**Comment by mjwitte** [comment]
Hmm, seems like that SP should be at the outlet of the HX coil system.  For that, we can't really control the water coil independently of the HX.

**Comment by EnergyArchmage** [comment]
The two error message are separate issues above.  The first is just a warning that the coil works better if the outlet node is the control node (but still allowed and runs).  the second error is severe if whatever the control node is doesn't have any setpoint on it.

**Comment by rraustad** [comment]
I ran with SP's at coil outlet node without any errors. Then moved both SP's to the HX outlet node. I only got the first warnings of sensor node not at water coil outlet, and no severe errors. So I must have done something wrong the first time.

Control with SP at water coil outlet node is good in both air loop and OA system (first pair of figures)
Control with SP at HX outlet is good in air loop but there appears to be a problem with this location in the OA system (second pair of figures). OR outlet node of coil system is higher than SP.
Then ran with SP at HX outlet in V8.7, much worse then, so this is improving (third pair of figures).

![hxassistedresults](https://user-images.githubusercontent.com/5778156/27925166-cd547da0-6251-11e7-92c9-bd8c846a26df.png)
